src/track.rs

```diff
@@ -344,85 +338,89 @@ impl Mp4Track {
-            let traf_idx = (sample_id - 1) / sample_sizes_count;
-            if let Some(trun) = &self.trafs[traf_idx as usize].trun {
-                if let Some(size) = trun.sample_sizes.get((sample_id - (sample_sizes_count * traf_idx)) as usize - 1) {
-                    Ok(*size)
-                } else {
-                    return Err(Error::EntryInTrunNotFound(
-                        self.track_id(),
-                        BoxType::TrunBox,
-                        sample_id,
-                    ));
-                }
-            } else {
-                return Err(Error::BoxInTrafNotFound(
-                    self.track_id(),
-                    BoxType::TrafBox,
-                ));
-            }
-        } else {
-            let stsz = &self.trak.mdia.minf.stbl.stsz;
-            if stsz.sample_size > 0 {
-                return Ok(stsz.sample_size);
-            }
-            if let Some(size) = stsz.sample_sizes.get(sample_id as usize - 1) {
-                Ok(*size)
-            } else {
-                return Err(Error::EntryInStblNotFound(
-                    self.track_id(),
-                    BoxType::StszBox,
-                    sample_id,
-                ));
-            }
-        }
-    }
-
-    fn total_sample_size(&self) -> u64 {
-        let stsz = &self.trak.mdia.minf.stbl.stsz;
-        if stsz.sample_size > 0 {
-            stsz.sample_size as u64 * self.sample_count() as u64
-        } else {
-            let mut total_size = 0;
-            for size in stsz.sample_sizes.iter() {
-                total_size += *size as u64;
-            }
-            total_size
-        }
-    }
-
-    fn sample_offset(&self, sample_id: u32) -> Result<u64> {
-        if self.trafs.len() > 0 {
-            let sample_sizes_count = self.sample_count() / self.trafs.len() as u32;
-            let traf_idx = (sample_id - 1) / sample_sizes_count;
-            Ok(self.trafs[(sample_id - (sample_sizes_count * traf_idx)) as usize].tfhd.base_data_offset as u64)
-        } else {
-            let stsc_index = self.stsc_index(sample_id);
-
-            let stsc = &self.trak.mdia.minf.stbl.stsc;
-            let stsc_entry = stsc.entries.get(stsc_index).unwrap();
-
-            let first_chunk = stsc_entry.first_chunk;
-            let first_sample = stsc_entry.first_sample;
-            let samples_per_chunk = stsc_entry.samples_per_chunk;
-
-            let chunk_id = first_chunk + (sample_id - first_sample) / samples_per_chunk;
-
-            let chunk_offset = self.chunk_offset(chunk_id)?;
-
-            let first_sample_in_chunk = sample_id - (sample_id - first_sample) % samples_per_chunk;
-
-            let mut sample_offset = 0;
-            for i in first_sample_in_chunk..sample_id {
-                sample_offset += self.sample_size(i)?;
-            }
-
-            Ok(chunk_offset + sample_offset as u64)
-        }
-    }
+    //fn sample_size(&self, sample_id: u32) -> Result<u32> {
```

Comment:
The biggest missing thing from this PR IMO is these functions being commented out.

Some of these are a bit tricky to update and make correct I believe, and are non trivial to make work on fragmented MP4s. `sample_size` is a good example, since it can be specified as a default both in the `mvex` box, the `tfhd` box, and overridden in the `trun` box.

There is also no guarantee that the sample data actually is in the same file as the header.

How do you feel about this?

Reply:
I see. I'm not really sure on the best approach yet to implementing these for fmp4 yet. It's part of the reason why I only implementing reading at first, ha.

Reply:
How do you feel about making a separate `ibmf` crate for reading and writing the ISO Base Media Format (that is, all the box readers and writers), then having the `mp4` crate depend on and use that, while providing higher level utilities?

We could then introduce other crates for reading and writing other variants of the IBMF if desired.

I still think it would make sense to have `fmp4` support be part of the `mp4` crate, since they just different variants that are mostly identical. I think in general it would take some API deign effort if the goal is to have a crate that can interact with a wide range of real world MP4 files.

Reply:
A separate `bmff` crate sounds pretty good. I actually started this project as a generic `bmff` reader in mind and it has changed around quite a bit. I like the idea of organizing the crates/packages by variants, with the variant-specific atom/boxes defined. Something like:
```
isobmff
├─ parsing
│     ├─ (f)mp4
|     |     └─ atom
│     ├─ qt
│     ├─ piff
│     └─ etc
└─ writing
      ├─ (f)mp4
      ├─ qt
      ├─ piff
      └─ etc


```
Or something like that. Kinda similar to how `mp4box.js` is organized:
https://github.com/gpac/mp4box.js/tree/master/src

However, my time is limited on this project at the moment, but open to submissions. And a restructure would probably need a development branch since it will be a large changeset. What do you think?